src/main/java/com/bwarelabs/BigTableToCosWriter.java

```diff
@@ -168,169 +180,126 @@ private String getThreadStartingKey(String tableName, String prefix, String maxP
-            try {
-                String currentStartRow = startRowKey;
-                List<CompletableFuture<Void>> batchFutures = new ArrayList<>();
-                boolean includeStartRow = !isCheckpointStart;
-
-                while (currentStartRow.compareTo(endRowKey) <= 0) {
-                    for (int i = 0; i < BATCH_LIMIT && currentStartRow.compareTo(endRowKey) <= 0; i++) {
-                        List<Result> batch = fetchBatchTx(tableName, currentStartRow, endRowKey, includeStartRow);
-                        if (batch.isEmpty()) {
-                            // since we use the last row key as the start row key for the next batch (non-inclusive interval)
-                            // the thread will hit a point where there are no more rows to process so we need to break out of the loop
-                            currentStartRow = endRowKey + "Z";
-                            break;
-                        }
-
-                        currentStartRow = Bytes.toString(batch.get(0).getRow());
-                        logger.info(String.format("[%s] Batch size: %s for startRow: %s and endRow: %s", Thread.currentThread().getName(), batch.size(), currentStartRow, endRowKey));
-                        String currentEndRow = Bytes.toString(batch.get(batch.size() - 1).getRow());
-
-                        List<CompletableFuture<Void>> tempBatchFutures = getUploadFutures(tableName, currentStartRow, currentEndRow, batch);
-                        batchFutures.addAll(tempBatchFutures);
-
-                        currentStartRow = currentEndRow;
-                        includeStartRow = false;
-                    }
-
-                    CompletableFuture<Void> currentBatch = CompletableFuture.allOf(batchFutures.toArray(new CompletableFuture[0]));
-                    currentBatch.join();
-                    updateCheckpoint(threadId, currentStartRow);
-                    batchFutures.clear();
-                }
-            } catch (Exception e) {
-                logger.severe(String.format("Error processing table range for %s - %s", tableName, e));
-                throw new RuntimeException(e);
-            }
-        }, executorService);
+    private void runTaskOnWorkerThread(int threadId, String tableName, String startRowKey, String endRowKey, boolean isCheckpointStart) {
+        CompletableFuture<Void> processingFuture = CompletableFuture.supplyAsync(() ->
+                        splitRangeAndChainUploads(threadId, tableName, startRowKey, endRowKey, !isCheckpointStart, new BatchCounter(), new ArrayList<>())
+                                .exceptionally(e -> {
+                                    logger.severe(String.format("Error processing table range for %s - %s", tableName, e));
+                                    throw new RuntimeException(e);
+                                })
+                , executorService).thenCompose(batchFuture -> batchFuture);
```

Comment:
there are 2 futures created in this block of code. 
One from creating a task on the executorService, and one returned by `splitRangeAndChainUploads`. 
This flattens / chains both of the futures so if i the `processingFuture` completes, it means that `splitRangeAndChainUploads` finished too. 
Otherwise, the one returned from spawning the task gets completed and the program exits without waiting for the actual upload to happen.